source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1679 @@
+One can also pass arguments by name.
+
+Example:
+
+```nix
+let
+  f = x: x.a;
+  v = { a = 1; };
+in
+f v
+```
+
+    1
+
+#### Multiple arguments
+
+Also known as “[curried] functions”.
+
+Nix functions take exactly one argument.
+Multiple arguments can be handled by nesting functions.
+
+Such a nested function can be used like a function that takes multiple arguments, but offers additional flexibility.
+
+[curried]: https://en.m.wikipedia.org/wiki/Currying
+
+Example:
+
+```nix
+x: y: x + y
+```
+
+    <LAMBDA>
+
+The above function is equivalent to
+
+
+```nix
+x: (y: x + y)
+```
+
+    <LAMBDA>
+
+This function takes one argument and returns another function `y: x + y` with `x` set to the value of the argument.
+
+Example:
+
+```nix
+let
+  f = x: y: x + y;
+in
+f 1
+```
+
+    <LAMBDA>
+
+
+Applying the function which results from `f 1` to another argument yields the inner body `x + y` (with `x` set to `1` and `y` set to the other argument), which can now be fully evaluated.
+
+```nix
+let
+  f = x: y: x + y;
+in
+f 1 2
+```
+
+    3
+
+<!-- TODO: exercise - assign the lambda a name and do something with it -->
+
+### Attribute set argument
+
+Also known as “keyword arguments” or “destructuring” .
+
+Nix functions can be declared to require an attribute set with specific structure as argument.
+
+This is denoted by listing the expected attribute names separated by commas (`,`) and enclosed in braces (`{ }`).
+
+Example:
+
+```nix
+{a, b}: a + b
+```
+
+    <LAMBDA>
+
+The argument defines the exact attributes that have to be in that set.
+Leaving out or passing additional attributes is an error.
+
+Example:
+
+```nix
+let
+  f = {a, b}: a + b;
+in
+f { a = 1; b = 2; }
+```
+
+    3
+
+Counter-example:
+
+```nix
+let
+  f = {a, b}: a + b;
+in
+f { a = 1; b = 2; c = 3; }
+```
+
+    error: 'f' at (string):2:7 called with unexpected argument 'c'
+
+           at «string»:4:1:
+
+                3| in
+                4| f { a = 1; b = 2; c = 3; }
+                 | ^
+                5|
+
+
+<!-- TODO: not the same as x: x.a + x.b (!!!!) -->
+
+#### Default values
+
+Also known as “default arguments”.
+
+Destructured arguments can have default values for attributes.
+
+This is denoted by separating the attribute name and its default value with a question mark (`?`).
+
+Attributes in the argument are not required if they have a default value.
+
+Example:
+
+```nix
+let
+  f = {a, b ? 0}: a + b;
+in
+f { a = 1; }
+```
+
+    1
+
+Example:
+
+```nix
+let
+  f = {a ? 0, b ? 0}: a + b;
+in
+f { } # empty attribute set
+```
+
+    0
+
+#### Additional attributes
+
+Additional attributes are allowed with an ellipsis (`...`):
+
+    {a, b, ...}: a + b
+
+Unlike in the previous counter-example, passing an argument that contains additional attributes is not an error.
+
+Example:
+
+```nix
+let
+  f = {a, b, ...}: a + b;
+in
+f { a = 1; b = 2; c = 3; }
+```
+
+    3
+
+### Named attribute set argument
```

Comment:
Note that while formals definitions can refer to each other (e.g. `{ a, b ? a * 2 }: ...`) they can not refer to the named pattern (e.g. `{ a, b ? args.a }@args: ...` is invalid)

Reply:
Important only for writing, not reading code.